source/server/filter_chain_manager_impl.h

```diff
@@ -234,34 +282,27 @@ class FilterChainManagerImpl : public Network::FilterChainManager,
   findFilterChainForSourceIpAndPort(const SourceIPsTrie& source_ips_trie,
                                     const Network::ConnectionSocket& socket) const;
 
+  // Duplicate the inherent factory context if any.
+  std::shared_ptr<Network::DrainableFilterChain>
+  findExistingFilterChain(const envoy::config::listener::v3::FilterChain& filter_chain_message);
+
   // Mapping of FilterChain's configured destination ports, IPs, server names, transport protocols
   // and application protocols, using structures defined above.
   DestinationPortsMap destination_ports_map_;
   const Network::Address::InstanceConstSharedPtr address_;
+  // This is the reference to a factory context where all the generations of listener share.
   Configuration::FactoryContext& parent_context_;
-  std::list<std::unique_ptr<Configuration::FilterChainFactoryContext>> factory_contexts_;
-};
-
-class FilterChainImpl : public Network::FilterChain {
-public:
-  FilterChainImpl(Network::TransportSocketFactoryPtr&& transport_socket_factory,
-                  std::vector<Network::FilterFactoryCb>&& filters_factory)
-      : transport_socket_factory_(std::move(transport_socket_factory)),
-        filters_factory_(std::move(filters_factory)) {}
-
-  // Network::FilterChain
-  const Network::TransportSocketFactory& transportSocketFactory() const override {
-    return *transport_socket_factory_;
-  }
+  std::list<std::shared_ptr<Configuration::FilterChainFactoryContext>> factory_contexts_;
 
-  const std::vector<Network::FilterFactoryCb>& networkFilterFactories() const override {
-    return filters_factory_;
-  }
+  // Reference to the previous generation of filter chain manager to share the filter chains.
+  // Caution: the pointer is legit only during warm up.
+  // TODO(lambdai): safer usage
```

Comment:
What does this mean? The origin_ concept seems pretty strange. Why does it need to be stored?

Reply:
I considered pointer because the original could be none. Let me see if I can remove

Reply:
I probably missed the context here due to the extraction. 
The origin filter chain manager need to be used to co-own the shared filter chains. It's not a pure transfer because the newer filter chain manager could be destroyed when another update replace the warming one.
The plan of safe usage is something like adding state_ to FilterChainManagerImpl, once the build phase is finished the origin cannot be accessed. Since the state transition is straightforward, I am not sure the plan is a overkill.

Reply:
Then please remove the TODO and make the comment more complete per what you said here.